tsh for Windows does not build with realtime backend

```
---> Building OSS binaries.
/usr/bin/make build/tsh
GOOS=windows GOARCH=`go env GOARCH` CGO_ENABLED=1 go build -tags pam -o build/tsh -ldflags -w -ldflags '-w -s' ./tool/tsh
# runtime/cgo
gcc: error: unrecognized command line option '-mthreads'; did you mean '-pthread'?
Makefile:80: recipe for target 'build/tsh' failed
make[2]: *** [build/tsh] Error 2
Makefile:62: recipe for target 'all' failed
make[1]: *** [all] Error 2
Makefile:116: recipe for target 'release' failed
make: *** [release] Error 2
make: *** [release-windows] Error 2
```

--- a/Makefile
+++ b/Makefile
@@ -107,7 +107,7 @@ endif
 clean:
 	@echo "---> Cleaning up OSS build artifacts."
 	rm -rf $(BUILDDIR)
-	go clean -cache
+	-go clean -cache
 	rm -rf $(GOPKGDIR)
 	rm -rf teleport
 	rm -rf *.gz

--- a/integration/integration_test.go
+++ b/integration/integration_test.go
@@ -1241,8 +1241,8 @@ func (s *IntSuite) TestHA(c *check.C) {
 	// try to execute an SSH command using the same old client  to Site-B
 	// "site-A" and "site-B" reverse tunnels are supposed to reconnect,
 	// and 'tc' (client) is also supposed to reconnect
-	for i := 0; i < 10; i++ {
-		time.Sleep(time.Millisecond * 50)
+	for i := 0; i < 30; i++ {
+		time.Sleep(1 * time.Second)
 		err = tc.SSH(context.TODO(), cmd, false)
 		if err == nil {
 			break

--- a/lib/auth/helpers.go
+++ b/lib/auth/helpers.go
@@ -28,7 +28,7 @@ import (
 	"github.com/gravitational/teleport"
 	authority "github.com/gravitational/teleport/lib/auth/testauthority"
 	"github.com/gravitational/teleport/lib/backend"
-	"github.com/gravitational/teleport/lib/backend/lite"
+	"github.com/gravitational/teleport/lib/backend/memory"
 	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/events"
 	"github.com/gravitational/teleport/lib/limiter"
@@ -112,10 +112,10 @@ func NewTestAuthServer(cfg TestAuthServerConfig) (*TestAuthServer, error) {
 		TestAuthServerConfig: cfg,
 	}
 	var err error
-	b, err := lite.NewWithConfig(context.Background(), lite.Config{
-		Path:             cfg.Dir,
-		PollStreamPeriod: 100 * time.Millisecond,
-		Clock:            cfg.Clock,
+	b, err := memory.New(memory.Config{
+		Context:   context.Background(),
+		Clock:     cfg.Clock,
+		EventsOff: false,
 	})
 	if err != nil {
 		return nil, trace.Wrap(err)

--- a/lib/auth/state.go
+++ b/lib/auth/state.go
@@ -24,9 +24,6 @@ import (
 
 	"github.com/gravitational/teleport"
 	"github.com/gravitational/teleport/lib/backend"
-	"github.com/gravitational/teleport/lib/backend/legacy"
-	"github.com/gravitational/teleport/lib/backend/legacy/dir"
-	"github.com/gravitational/teleport/lib/backend/lite"
 	"github.com/gravitational/teleport/lib/services"
 	"github.com/gravitational/teleport/lib/utils"
 
@@ -38,26 +35,6 @@ import (
 // and keeps local process credentials - x509 and SSH certs and keys.
 type ProcessStorage struct {
 	backend.Backend
-}
-
-// NewProcessStorage returns a new instance of the process storage.
-func NewProcessStorage(ctx context.Context, path string) (*ProcessStorage, error) {
-	if path == "" {
-		return nil, trace.BadParameter("missing parameter path")
-	}
-	litebk, err := lite.NewWithConfig(ctx, lite.Config{Path: path, EventsOff: true})
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	// Import storage data
-	err = legacy.Import(ctx, litebk, func() (legacy.Exporter, error) {
-		return dir.New(legacy.Params{"path": path})
-	})
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
-	return &ProcessStorage{Backend: litebk}, nil
 }
 
 // Close closes all resources used by process storage backend.

--- a/lib/auth/state_unix.go
+++ b/lib/auth/state_unix.go
@@ -1,0 +1,50 @@
+// +build !windows
+
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+
+	"github.com/gravitational/teleport/lib/backend/legacy"
+	"github.com/gravitational/teleport/lib/backend/legacy/dir"
+	"github.com/gravitational/teleport/lib/backend/lite"
+
+	"github.com/gravitational/trace"
+)
+
+// NewProcessStorage returns a new instance of the process storage.
+func NewProcessStorage(ctx context.Context, path string) (*ProcessStorage, error) {
+	if path == "" {
+		return nil, trace.BadParameter("missing parameter path")
+	}
+
+	litebk, err := lite.NewWithConfig(ctx, lite.Config{Path: path, EventsOff: true})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// Import storage data
+	err = legacy.Import(ctx, litebk, func() (legacy.Exporter, error) {
+		return dir.New(legacy.Params{"path": path})
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &ProcessStorage{Backend: litebk}, nil
+}

--- a/lib/auth/state_windows.go
+++ b/lib/auth/state_windows.go
@@ -1,0 +1,40 @@
+// +build windows
+
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+
+	"github.com/gravitational/teleport/lib/backend/memory"
+
+	"github.com/gravitational/trace"
+)
+
+// NewProcessStorage returns a new instance of the process storage.
+func NewProcessStorage(ctx context.Context, path string) (*ProcessStorage, error) {
+	m, err := memory.New(memory.Config{
+		Context:   ctx,
+		EventsOff: true,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &ProcessStorage{Backend: m}, nil
+}

--- a/lib/backend/memory/memory.go
+++ b/lib/backend/memory/memory.go
@@ -155,10 +155,14 @@ func (m *Memory) Create(ctx context.Context, i backend.Item) (*backend.Lease, er
 	if m.tree.Get(&btreeItem{Item: i}) != nil {
 		return nil, trace.AlreadyExists("key %q already exists", string(i.Key))
 	}
-	m.processEvent(backend.Event{
+	event := backend.Event{
 		Type: backend.OpPut,
 		Item: i,
-	})
+	}
+	m.processEvent(event)
+	if !m.EventsOff {
+		m.buf.Push(event)
+	}
 	return m.newLease(i), nil
 }
 
@@ -289,10 +293,14 @@ func (m *Memory) DeleteRange(ctx context.Context, startKey, endKey []byte) error
 	m.removeExpired()
 	re := m.getRange(ctx, startKey, endKey, backend.NoLimit)
 	for _, item := range re.Items {
-		m.processEvent(backend.Event{
+		event := backend.Event{
 			Type: backend.OpDelete,
 			Item: item,
-		})
+		}
+		m.processEvent(event)
+		if !m.EventsOff {
+			m.buf.Push(event)
+		}
 	}
 	return nil
 }

--- a/lib/events/auditlog.go
+++ b/lib/events/auditlog.go
@@ -30,7 +30,6 @@ import (
 	"sort"
 	"strings"
 	"sync"
-	"syscall"
 	"time"
 
 	"github.com/gravitational/teleport"
@@ -998,7 +997,7 @@ func (l *AuditLog) periodicSpaceMonitor() {
 		case <-ticker.C:
 			// Find out what percentage of disk space is used. If the syscall fails,
 			// emit that to prometheus as well.
-			usedPercent, err := percentUsed(l.DataDir)
+			usedPercent, err := utils.PercentUsed(l.DataDir)
 			if err != nil {
 				auditFailedDisk.Inc()
 				log.Warnf("Disk space monitoring failed: %v.", err)
@@ -1016,17 +1015,4 @@ func (l *AuditLog) periodicSpaceMonitor() {
 			return
 		}
 	}
-}
-
-// percentUsed returns percentage of disk space used. The percentage of disk
-// space used is calculated from (total blocks - free blocks)/total blocks.
-// The value is rounded to the nearest whole integer.
-func percentUsed(path string) (float64, error) {
-	var stat syscall.Statfs_t
-	err := syscall.Statfs(path, &stat)
-	if err != nil {
-		return 0, trace.Wrap(err)
-	}
-	ratio := float64(stat.Blocks-stat.Bfree) / float64(stat.Blocks)
-	return math.Round(ratio * 100), nil
 }

--- a/lib/services/suite/suite.go
+++ b/lib/services/suite/suite.go
@@ -1423,6 +1423,7 @@ func eventsTestKinds(tests []eventTest) []services.WatchKind {
 	return out
 }
 
+// ExpectResource expects a Put event of a certain resource
 func ExpectResource(c *check.C, w services.Watcher, timeout time.Duration, resource services.Resource) {
 	timeoutC := time.After(timeout)
 waitLoop:
@@ -1451,13 +1452,14 @@ waitLoop:
 	}
 }
 
+// ExpectDeleteResource expects a delete event of a certain kind
 func ExpectDeleteResource(c *check.C, w services.Watcher, timeout time.Duration, resource services.Resource) {
 	timeoutC := time.After(timeout)
 waitLoop:
 	for {
 		select {
 		case <-timeoutC:
-			c.Fatalf("Timeout waiting for event")
+			c.Fatalf("Timeout waiting for delete resource %v", resource)
 		case <-w.Done():
 			c.Fatalf("Watcher exited with error %v", w.Error())
 		case event := <-w.Events():

--- a/lib/utils/disk.go
+++ b/lib/utils/disk.go
@@ -1,0 +1,38 @@
+// +build !windows
+
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"syscall"
+
+	"github.com/gravitational/trace"
+)
+
+// PercentUsed returns percentage of disk space used. The percentage of disk
+// space used is calculated from (total blocks - free blocks)/total blocks.
+// The value is rounded to the nearest whole integer.
+func PercentUsed(path string) (float64, error) {
+	var stat syscall.Statfs_t
+	err := syscall.Statfs(path, &stat)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	ratio := float64(stat.Blocks-stat.Bfree) / float64(stat.Blocks)
+	return Round(ratio * 100), nil
+}

--- a/lib/utils/disk_windows.go
+++ b/lib/utils/disk_windows.go
@@ -1,0 +1,26 @@
+// +build windows
+
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import "github.com/gravitational/trace"
+
+// PercentUsed not supported on Windows.
+func PercentUsed(path string) (float64, error) {
+	return 0.0, trace.BadParameter("disk usage not supported on Windows")
+}

--- a/lib/utils/round.go
+++ b/lib/utils/round.go
@@ -1,0 +1,58 @@
+// Copyright 2009 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package utils
+
+import (
+	"math"
+)
+
+const (
+	uvone    = 0x3FF0000000000000
+	mask     = 0x7FF
+	shift    = 64 - 11 - 1
+	bias     = 1023
+	signMask = 1 << 63
+	fracMask = 1<<shift - 1
+)
+
+// Round returns the nearest integer, rounding half away from zero.
+//
+// Special cases are:
+//  Round(±0) = ±0
+//  Round(±Inf) = ±Inf
+//  Round(NaN) = NaN
+//
+// Note: Copied from Go standard library to support Go 1.9.7 releases. This
+// function was added in the standard library in Go 1.10.
+func Round(x float64) float64 {
+	// Round is a faster implementation of:
+	//
+	// func Round(x float64) float64 {
+	//   t := Trunc(x)
+	//   if Abs(x-t) >= 0.5 {
+	//     return t + Copysign(1, x)
+	//   }
+	//   return t
+	// }
+	bits := math.Float64bits(x)
+	e := uint(bits>>shift) & mask
+	if e < bias {
+		// Round abs(x) < 1 including denormals.
+		bits &= signMask // +-0
+		if e == bias-1 {
+			bits |= uvone // +-1
+		}
+	} else if e < bias+shift {
+		// Round any abs(x) >= 1 containing a fractional component [0,1).
+		//
+		// Numbers with larger exponents are returned unchanged since they
+		// must be either an integer, infinity, or NaN.
+		const half = 1 << (shift - 1)
+		e -= bias
+		bits += half >> e
+		bits &^= fracMask >> e
+	}
+	return math.Float64frombits(bits)
+}

